Formula/llvm.rb

```diff
@@ -347,8 +348,8 @@ def install
         ldflags = %W[
           -L#{libstdcxx_dir}
           -Wl,-rpath,#{libstdcxx_dir}
-          -L#{HOMEBREW_PREFIX}/lib
-          -Wl,-rpath,#{HOMEBREW_PREFIX}/lib
+          -L#{Formula["glibc"].opt_lib}
+          -Wl,-rpath,#{Formula["glibc"].opt_lib}
         ]
```

Comment:
I'm wondering about these flags on Linux. Why not just add them once and for all in the beginning and use them in each build?

Also, shouldn't we be adding the `glibc` flags only if `glibc` is installed? Or is that there so that it will use Homebrew `glibc` when needed?

Reply:
Also, why not use `CMAKE_INSTALL_RPATH`?

Reply:
> Also, shouldn't we be adding the `glibc` flags only if `glibc` is installed? Or is that there so that it will use Homebrew `glibc` when needed?

It's there so that Homebrew `glibc` will be used when needed - it will have no effect if no `glibc` is installed.

> I'm wondering about these flags on Linux. Why not just add them once and for all in the beginning and use them in each build?
> Also, why not use `CMAKE_INSTALL_RPATH`?

These both sound like good ideas though I'm inclined to say at the moment that we should do the bare minimum needed to fix this right now since LLVM is acutely broken for users with `gcc@5` installed.

One complexity with sharing these arguments between the PGO build and final build is that we are using `extra_args` for the PGO build and `args` for the final build.  We could store the `CMAKE_INSTALL_RPATH` and `CMAKE_LIBRARY_PATH` flags in their own array and append it to both `extra_args` and `args` on Linux.  That would simplify things quite a bit, though I'm not sure if that's exactly what you had in mind.

Reply:
Yep, that sounds good. Something like this?

```diff
diff --git a/Formula/llvm.rb b/Formula/llvm.rb
index bc4ac7bc8f7..8cd83061e21 100644
--- a/Formula/llvm.rb
+++ b/Formula/llvm.rb
@@ -194,6 +194,8 @@ class Llvm < Formula
     args << "-DBUILTINS_CMAKE_ARGS=#{builtins_cmake_args.join(";")}" if builtins_cmake_args.present?
 
     llvmpath = buildpath/"llvm"
+    gcc_version = Formula["gcc"].version.major
+
     if pgo_build
       # We build LLVM a few times first for optimisations. See
       # https://github.com/Homebrew/homebrew-core/issues/77975
@@ -214,6 +216,20 @@ class Llvm < Formula
         # Make sure CMake doesn't try to pass C++-only flags to C compiler.
         extra_args << "-DCMAKE_C_COMPILER=#{ENV.cc}"
         extra_args << "-DCMAKE_CXX_COMPILER=#{ENV.cxx}"
+
+        linux_library_paths = [
+          rpath,
+          Formula["gcc"].opt_lib/"gcc"/gcc_version, # libstdc++
+          Formula["glibc"].opt_lib,
+        ]
+
+        # Link to libstdc++ for brewed GCC rather than the host GCC which is too old.
+        extra_args << "-DCMAKE_INSTALL_RPATH=#{linux_library_paths.join(";")}"
+        extra_args << "-DCMAKE_SYSTEM_LIBRARY_PATH=#{linux_library_paths.join(";")}"
+
+        # We need this for the installaed toolchain too.
+        args << "-DCMAKE_INSTALL_RPATH=#{linux_library_paths.join(";")}"
+        args << "-DCMAKE_SYSTEM_LIBRARY_PATH=#{linux_library_paths.join(";")}"
       end
 
       cflags = ENV.cflags&.split || []
@@ -252,27 +268,19 @@ class Llvm < Formula
 
         extra_args.reject! { |s| s["CMAKE_CXX_FLAGS"] }
         extra_args << "-DCMAKE_CXX_FLAGS=#{cxxflags.join(" ")}"
-      end
-
-      # On Linux, our just-built Clang needs a little help finding C++ headers,
-      # since we did not build libc++, and we are using the libstdc++ headers instead.
-      # We also need to make sure it links to libstdc++ for brewed GCC rather than
-      # the host GCC which is too old.
-      unless OS.mac?
-        gcc_version = Formula["gcc"].version.major
-
+      elsif !OS.mac?
+        # On Linux, our just-built Clang needs a little help finding C++ headers,
+        # since we did not build libc++, and we are using the libstdc++ headers instead.
         cxxflags << "-isystem#{Formula["gcc"].opt_include}/c++/#{gcc_version}"
         cxxflags << "-isystem#{Formula["gcc"].opt_include}/c++/#{gcc_version}/x86_64-pc-linux-gnu"
         cxxflags << "-nostdinc++"
 
+        # Make sure stage1 clang can find headers for other formulae on Linux.
+        cxxflags << "-isystem#{HOMEBREW_PREFIX}/include"
+
         extra_args.reject! { |s| s["CMAKE_CXX_FLAGS"] }
-        extra_args.reject! { |s| s["CMAKE_C_COMPILER"] }
-        extra_args.reject! { |s| s["CMAKE_CXX_COMPILER"] }
+        extra_args.reject! { |s| s[/CMAKE_C(XX)?_COMPILER/] }
         extra_args << "-DCMAKE_CXX_FLAGS=#{cxxflags.join(" ")}"
-
-        libstdcxx_dir = Formula["gcc"].opt_lib/"gcc"/gcc_version
-        extra_args << "-DCMAKE_EXE_LINKER_FLAGS=-L#{libstdcxx_dir} -Wl,-rpath,#{libstdcxx_dir}"
-        extra_args << "-DCMAKE_SHARED_LINKER_FLAGS=-L#{libstdcxx_dir} -Wl,-rpath,#{libstdcxx_dir}"
       end
 
       # Next, build an instrumented stage2 compiler
@@ -335,24 +343,9 @@ class Llvm < Formula
       # Silence some warnings
       cflags << "-Wno-backend-plugin"
       cxxflags << "-Wno-backend-plugin"
-      # Make sure stage1 clang can find headers for other formulae on Linux.
-      cxxflags << "-isystem#{HOMEBREW_PREFIX}/include" unless OS.mac?
 
       args << "-DCMAKE_C_FLAGS=#{cflags.join(" ")}"
       args << "-DCMAKE_CXX_FLAGS=#{cxxflags.join(" ")}"
-
-      # Add linker flags on Linux so that stage1 clang can find libstdc++ from brewed GCC and all of the other
-      # brewed libraries that need to be linked.
-      unless OS.mac?
-        ldflags = %W[
-          -L#{libstdcxx_dir}
-          -Wl,-rpath,#{libstdcxx_dir}
-          -L#{HOMEBREW_PREFIX}/lib
-          -Wl,-rpath,#{HOMEBREW_PREFIX}/lib
-        ]
-        args << "-DCMAKE_EXE_LINKER_FLAGS=#{ldflags.join(" ")}"
-        args << "-DCMAKE_SHARED_LINKER_FLAGS=#{ldflags.join(" ")}"
-      end
     end
 
     # Now, we can build.
```

I'm actually a little bit tempted to replace all of the `-isystem` flags with `CMAKE_SYSTEM_INCLUDE_PATH`, but I'm a little less clear that that does the same thing.

Reply:
Some of the flags are passed a bit earlier than they're needed, but I find it a lot easier to understand what's going on after these changes (since there are much fewer `OS` conditionals).

Reply:
I think I will try using `CMAKE_SYSTEM_INCLUDE_PATH` as it doesn't really matter if those include paths use `-I` or `-isystem` - as I understand it the main difference between those two flags is the search order and we're using `nostdinc++` so it won't be looking for system `libstdc++` headers anyway.  That would let us further confine all but one of the Linux specific changes to a single block at the beginning.  

As you've noted it's technically redundant for us to pass the include and linker flags during the stage 1 build because the GCC spec and compiler shim already do this for us, but that shouldn't break anything and it certainly makes this whole thing a lot simpler to follow.

Reply:
I've pushed the changes you suggested along with using `CMAKE_SYSTEM_INCLUDE_PATH` instead of all those `isystem` CXXFLAGS.  I think I will requeue CI to run again on a regular 6 hour GitHub Actions runner.  Although it may not quite finish, I'm guessing it's at least getting to the final build so if we broke something during the PGO part it will fail earlier.